// hltv/lib/endpoints/getMatches.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMatches = void 0;
var querystring_1 = require("querystring");
var parsing_1 = require("../utils/parsing");
var mappers_1 = require("../utils/mappers");
var checkForRateLimiting_1 = require("../utils/checkForRateLimiting");
var getMatches = function (config) { return function (_a) {
    var _b = _a === void 0 ? {} : _a, eventID = _b.eventID, eventType = _b.eventType, filter = _b.filter;
    return __awaiter(void 0, void 0, void 0, function () {
        var query, $, liveMatches, upcomingMatches;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    query = querystring_1.stringify(__assign(__assign(__assign({}, (eventID ? { event: eventID } : {})), (eventType ? { eventType: eventType } : {})), (filter ? { predefinedFilter: filter } : {})));
                    return [4, mappers_1.fetchPage(config.hltvUrl + "/matches?" + query, config.loadPage)];
                case 1:
                    $ = _c.sent();
                    checkForRateLimiting_1.checkForRateLimiting($);
                    liveMatches = mappers_1.toArray($('.liveMatch-container')).map(function (matchEl) {
                        var id = Number(matchEl.find('.a-reset').attr('href').split('/')[2]);
                        var teamNameEls = matchEl.find('.matchTeamName');
                        var stars = 5 - matchEl.find('.matchRating i.faded').length;
                        var team1 = {
                            name: teamNameEls.first().text(),
                            id: Number(matchEl.attr('team1')) || undefined
                        };
                        var team2 = {
                            name: teamNameEls.last().text(),
                            id: Number(matchEl.attr('team2')) || undefined
                        };
                        var format = matchEl.find('.matchMeta').text();
                        var event = {
                            name: matchEl.find('.matchEventLogo').attr('title'),
                            id: Number(parsing_1.popSlashSource(matchEl.find('.matchEventLogo')).split('.')[0]) || undefined
                        };
                        return { id: id, team1: team1, team2: team2, event: event, format: format, stars: stars, live: true };
                    });
                    upcomingMatches = mappers_1.toArray($('.upcomingMatch')).map(function (matchEl) {
                        var link = matchEl.find('.a-reset');
                        var id = Number(link.attr('href').split('/')[2]);
                        var date = Number(matchEl.find('.matchTime').attr('data-unix')) || undefined;
                        var title = matchEl.find('.matchInfoEmpty').text() || undefined;
                        var stars = matchEl.find('.matchRating i:not(.faded)').length;
                        var format = matchEl.find('.matchMeta').text();
                        var event;
                        var team1;
                        var team2;
                        if (!title) {
                            team1 = {
                                name: matchEl.find('.team1 .matchTeamName').text(),
                                id: Number(matchEl.attr('team1')) || undefined
                            };
                            team2 = {
                                name: matchEl.find('.team2 .matchTeamName').text() ||
                                    matchEl.find('.team2 .team').text(),
                                id: Number(matchEl.attr('team2')) || undefined
                            };
                            event = {
                                name: matchEl.find('.matchEventLogo').attr('alt'),
                                id: Number(parsing_1.popSlashSource(matchEl.find('.matchEventLogo')).split('.')[0]) || undefined
                            };
                        }
                        return {
                            id: id,
                            date: date,
                            team1: team1,
                            team2: team2,
                            format: format,
                            title: title,
                            event: event,
                            stars: stars,
                            live: false
                        };
                    });
                    return [2, __spreadArray(__spreadArray([], liveMatches), upcomingMatches)];
            }
        });
    });
}; };
exports.getMatches = getMatches;
